File: apps/agents/src/open-canvas/nodes/generateTitle.ts
import { LangGraphRunnableConfig } from "@langchain/langgraph";
import { Client } from "@langchain/langgraph-sdk";
import { OpenCanvasGraphAnnotation } from "../state.js";

export const generateTitleNode = async (
  state: typeof OpenCanvasGraphAnnotation.State,
  config: LangGraphRunnableConfig
) => {
  if (state.messages.length > 2) {
    // Skip if it's not first human ai conversation. Should never occur in practice
    // due to the conditional edge which is called before this node.
    return {};
  }

  try {
    const langGraphClient = new Client({
      apiUrl: `http://localhost:${process.env.PORT}`,
    });

    const titleInput = {
      messages: state.messages,
      artifact: state.artifact,
    };
    const titleConfig = {
      configurable: {
        open_canvas_thread_id: config.configurable?.thread_id,
      },
    };

    // Create a new thread for title generation
    const newThread = await langGraphClient.threads.create();

    // Create a new title generation run in the background
    await langGraphClient.runs.create(newThread.thread_id, "thread_title", {
      input: titleInput,
      config: titleConfig,
      multitaskStrategy: "enqueue",
      afterSeconds: 0,
    });
  } catch (e) {
    console.error("Failed to call generate title graph\n\n", e);
  }

  return {};
};

File: agents-py/agents/src/open_canvas/nodes/generate_title.py
import os
from typing import Dict, Any
from langgraph_sdk import get_client
from langchain_core.runnables import RunnableConfig
from agents.src.open_canvas.state import OpenCanvasGraphState

async def generate_title(
    state: OpenCanvasGraphState,
    config: RunnableConfig
) -> Dict[str, Any]:
    if len(state.messages) > 2:
        # Skip if not first human-AI conversation
        return {}

    try:
        lang_graph_client = get_client(url=f"http://localhost:{os.getenv('PORT', '8000')}")

        title_input = {
            "messages": state.messages,
            "artifact": state.artifact.dict() if state.artifact else None
        }
        
        title_config = {
            "configurable": {
                "open_canvas_thread_id": config.get("configurable", {}).get("thread_id")
            }
        }

        # Create new thread for title generation
        new_thread = await lang_graph_client.threads.create()
        
        # Start title generation run in background
        await lang_graph_client.runs.create(
            thread_id=new_thread.thread_id,
            app_id="thread_title",
            input=title_input,
            config=title_config,
            multitask_strategy="enqueue",
            after_seconds=0
        )
        
    except Exception as e:
        print(f"Failed to call generate title graph\n\n{e}")

    return {} 